### ISSUE TITLE
defun target finder matches too often

### ISSUE THREAD

**Issue by minad** [oantolin/embark#422]
In our action keymaps we assume that defuns are expressions. However defuns match something else in text modes. Does it make sense to refine this, distinguish general defun-like targets and elisp-defuns which are expressions? We could also consider to restrict the defun target to elisp buffers only. Related #421.

**Comment by minad** [comment]
@oantolin What do you think about this? After we got #421 I am less concernced with this issue. My only remaining gripe is: 

> In our action keymaps we assume that defuns are expressions.

**Comment by oantolin** [comment]
Also, most of the actions in the defun map only makes sense in Emacs Lisp buffers. So I think we probably should separate the defun target into several targets, but I think I need to through the list of actions to see which make sense in which modes to decide whether to just split it into Emacs Lisp vs everything else, or into 3: Emacs Lisp, other prog modes, non prog modes.

**Comment by minad** [comment]
Right. Three different modes make sense: elisp (possibly deriving from prog), prog and text.

However from a user perspective defuns in text modes don't make much sense. They are abused as some kind of navigational unit. Therefore one could also consider to drop support for text modes.

Also note that it is good to retain defun support in text modes but in the sense that the actions are acting on elisp defuns/expressions (org code snippets, literate code). I must admit that back then when I proposed defun/expression support I didn't really get this and did not consider that defun finders are overwritten and weirdly redefined in modes where defuns are not really defuns.

**Comment by minad** [comment]
With this one ends up with:

* elisp defun in elisp
* elisp defun in text (literate, org src blocks etc)
* prog defuns (mode specific). This is a bit heretical but on could also consider dropping support since there are not many useful general actions. I am looking forward to treesitter which will change this, but then one will get much more general treesitter targets. There seems to be a lot of movement recently with regards to treesitter. We should keep it in mind.
* text defuns (nosense, at least the name :-P). Maybe drop support for this?
